File: attached_assets/_extracted/Competency-Platform-main/src/server/api/routers/kpi.ts
import { z } from "zod";
import { TRPCError } from "@trpc/server";

import { createTRPCRouter, adminProcedure } from "@/server/api/trpc";
import { idSchema } from "@/lib/validators/common";
import { kpiSaveSchema } from "@/lib/validators/kpi";
import { isAiEnabled, generateKpis } from "@/server/services/ai";

export const kpiRouter = createTRPCRouter({
  /** Whether AI generation is configured (no secrets exposed). */
  aiEnabled: adminProcedure.query(() => isAiEnabled()),

  /** Jobs with a summary of their KPI coverage. */
  list: adminProcedure.query(async ({ ctx }) => {
    const jobs = await ctx.db.job.findMany({
      select: {
        id: true,
        name: true,
        department: { select: { name: true } },
        _count: { select: { competencies: true } },
        kpiSet: {
          select: {
            id: true,
            isAiGenerated: true,
            groups: { select: { _count: { select: { kpis: true } } } },
          },
        },
      },
      orderBy: { name: "asc" },
    });
    return jobs.map((j) => ({
      id: j.id,
      name: j.name,
      department: j.department?.name ?? null,
      competencyCount: j._count.competencies,
      hasKpis: Boolean(j.kpiSet),
      isAiGenerated: j.kpiSet?.isAiGenerated ?? false,
      kpiCount: j.kpiSet?.groups.reduce((sum, g) => sum + g._count.kpis, 0) ?? 0,
    }));
  }),

  /** KPI set for a job + the job's competencies (for the builder). */
  get: adminProcedure.input(z.object({ jobId: idSchema })).query(async ({ ctx, input }) => {
    const job = await ctx.db.job.findUnique({
      where: { id: input.jobId },
      select: {
        id: true,
        name: true,
        competencies: { select: { competency: { select: { name: true } } } },
        kpiSet: {
          select: {
            summary: true,
            isAiGenerated: true,
            groups: {
              orderBy: { order: "asc" },
              select: {
                competencyName: true,
                compType: true,
                kpis: {
                  orderBy: { order: "asc" },
                  select: { name: true, description: true, measure: true, target: true, frequency: true, weight: true },
                },
              },
            },
          },
        },
      },
    });
    if (!job) throw new TRPCError({ code: "NOT_FOUND" });
    return {
      job: { id: job.id, name: job.name, competencies: job.competencies.map((c) => c.competency.name) },
      kpiSet: job.kpiSet,
    };
  }),

  /** Create or replace a job's KPI set. */
  save: adminProcedure.input(kpiSaveSchema).mutation(async ({ ctx, input }) => {
    const groupsData = input.groups.map((g, gi) => ({
      competencyName: g.competencyName,
      compType: g.compType ?? null,
      order: gi,
      kpis: {
        create: g.kpis.map((k, ki) => ({
          name: k.name,
          description: k.description ?? null,
          measure: k.measure ?? null,
          target: k.target ?? null,
          frequency: k.frequency ?? null,
          weight: k.weight ?? null,
          order: ki,
        })),
      },
    }));

    return ctx.db.$transaction(async (tx) => {
      const existing = await tx.kpiSet.findUnique({ where: { jobId: input.jobId }, select: { id: true } });
      if (existing) {
        await tx.kpiGroup.deleteMany({ where: { kpiSetId: existing.id } });
        return tx.kpiSet.update({
          where: { id: existing.id },
          data: { summary: input.summary ?? null, isAiGenerated: input.isAiGenerated, groups: { create: groupsData } },
          select: { id: true },
        });
      }
      return tx.kpiSet.create({
        data: {
          jobId: input.jobId,
          summary: input.summary ?? null,
          isAiGenerated: input.isAiGenerated,
          groups: { create: groupsData },
        },
        select: { id: true },
      });
    });
  }),

  /** AI-generate KPIs for a job (not persisted — returned for review). */
  generate: adminProcedure.input(z.object({ jobId: idSchema })).mutation(async ({ ctx, input }) => {
    const job = await ctx.db.job.findUnique({
      where: { id: input.jobId },
      select: { name: true, competencies: { select: { competency: { select: { name: true } } } } },
    });
    if (!job) throw new TRPCError({ code: "NOT_FOUND" });
    return generateKpis({ name: job.name, competencies: job.competencies.map((c) => c.competency.name) });
  }),

  delete: adminProcedure.input(z.object({ jobId: idSchema })).mutation(async ({ ctx, input }) => {
    await ctx.db.kpiSet.deleteMany({ where: { jobId: input.jobId } });
    return { jobId: input.jobId };
  }),
});

File: artifacts/api-server/internal/store/kpi.go
package store

import (
	"context"
	"errors"

	"github.com/jackc/pgx/v4"

	"competency/internal/domain"
)

// JobKpiStatus describes a job's KPI coverage for the KPI list page.
type JobKpiStatus struct {
	JobID        string  `json:"jobId"`
	JobName      string  `json:"jobName"`
	DepartmentName *string `json:"departmentName"`
	HasKpiSet    bool    `json:"hasKpiSet"`
	KpiCount     int     `json:"kpiCount"`
}

// ListJobKpiStatus lists jobs with their KPI coverage status.
func (s *Store) ListJobKpiStatus(ctx context.Context) ([]JobKpiStatus, error) {
	rows, err := s.pool.Query(ctx, `
		SELECT j.id, j.name, d.name,
			(ks.id IS NOT NULL) AS has_set,
			COALESCE((SELECT count(*) FROM "Kpi" k JOIN "KpiGroup" kg ON kg.id = k."kpiGroupId" WHERE kg."kpiSetId" = ks.id), 0)
		FROM "Job" j
		LEFT JOIN "Department" d ON d.id = j."departmentId"
		LEFT JOIN "KpiSet" ks ON ks."jobId" = j.id
		ORDER BY j.name`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := []JobKpiStatus{}
	for rows.Next() {
		var s JobKpiStatus
		if err := rows.Scan(&s.JobID, &s.JobName, &s.DepartmentName, &s.HasKpiSet, &s.KpiCount); err != nil {
			return nil, err
		}
		out = append(out, s)
	}
	return out, rows.Err()
}

// KpiSetByJob returns the KPI set (with groups and kpis) for a job, or nil.
func (s *Store) KpiSetByJob(ctx context.Context, jobID string) (*domain.KpiSet, error) {
	var ks domain.KpiSet
	row := s.pool.QueryRow(ctx, `SELECT id, "jobId", summary, "isAiGenerated" FROM "KpiSet" WHERE "jobId"=$1`, jobID)
	if err := row.Scan(&ks.ID, &ks.JobID, &ks.Summary, &ks.IsAiGenerated); err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return nil, nil
		}
		return nil, err
	}
	ks.Groups = []domain.KpiGroup{}

	grows, err := s.pool.Query(ctx, `
		SELECT id, "kpiSetId", "competencyName", "compType", "order" FROM "KpiGroup" WHERE "kpiSetId"=$1 ORDER BY "order"`, ks.ID)
	if err != nil {
		return nil, err
	}
	defer grows.Close()
	groupIdx := map[string]int{}
	for grows.Next() {
		var g domain.KpiGroup
		if err := grows.Scan(&g.ID, &g.KpiSetID, &g.CompetencyName, &g.CompType, &g.Order); err != nil {
			return nil, err
		}
		g.Kpis = []domain.Kpi{}
		groupIdx[g.ID] = len(ks.Groups)
		ks.Groups = append(ks.Groups, g)
	}
	if err := grows.Err(); err != nil {
		return nil, err
	}

	krows, err := s.pool.Query(ctx, `
		SELECT k.id, k."kpiGroupId", k.name, k.description, k.measure, k.target, k.frequency, k.weight, k."order"
		FROM "Kpi" k JOIN "KpiGroup" kg ON kg.id = k."kpiGroupId"
		WHERE kg."kpiSetId"=$1 ORDER BY k."order"`, ks.ID)
	if err != nil {
		return nil, err
	}
	defer krows.Close()
	for krows.Next() {
		var k domain.Kpi
		if err := krows.Scan(&k.ID, &k.KpiGroupID, &k.Name, &k.Description, &k.Measure, &k.Target, &k.Frequency, &k.Weight, &k.Order); err != nil {
			return nil, err
		}
		if idx, ok := groupIdx[k.KpiGroupID]; ok {
			ks.Groups[idx].Kpis = append(ks.Groups[idx].Kpis, k)
		}
	}
	return &ks, krows.Err()
}

// KpiSetInput is the save payload.
type KpiSetInput struct {
	Summary       *string         `json:"summary"`
	IsAiGenerated bool            `json:"isAiGenerated"`
	Groups        []KpiGroupInput `json:"groups"`
}

// KpiGroupInput is a nested group.
type KpiGroupInput struct {
	CompetencyName string     `json:"competencyName"`
	CompType       *string    `json:"compType"`
	Kpis           []KpiInput `json:"kpis"`
}

// KpiInput is a nested kpi.
type KpiInput struct {
	Name        string  `json:"name"`
	Description *string `json:"description"`
	Measure     *string `json:"measure"`
	Target      *string `json:"target"`
	Frequency   *string `json:"frequency"`
	Weight      *string `json:"weight"`
}

// SaveKpiSet creates or replaces the entire KPI set for a job.
func (s *Store) SaveKpiSet(ctx context.Context, jobID string, in KpiSetInput) error {
	tx, err := s.pool.Begin(ctx)
	if err != nil {
		return err
	}
	defer tx.Rollback(ctx)

	if _, err := tx.Exec(ctx, `DELETE FROM "KpiSet" WHERE "jobId"=$1`, jobID); err != nil {
		return err
	}
	setID := NewID()
	if _, err := tx.Exec(ctx, `
		INSERT INTO "KpiSet" (id, "jobId", summary, "isAiGenerated", "createdAt", "updatedAt")
		VALUES ($1,$2,$3,$4, now(), now())`, setID, jobID, in.Summary, in.IsAiGenerated); err != nil {
		return err
	}
	for gi, g := range in.Groups {
		groupID := NewID()
		if _, err := tx.Exec(ctx, `
			INSERT INTO "KpiGroup" (id, "kpiSetId", "competencyName", "compType", "order")
			VALUES ($1,$2,$3,$4,$5)`, groupID, setID, g.CompetencyName, g.CompType, gi); err != nil {
			return err
		}
		for ki, k := range g.Kpis {
			if _, err := tx.Exec(ctx, `
				INSERT INTO "Kpi" (id, "kpiGroupId", name, description, measure, target, frequency, weight, "order")
				VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9)`,
				NewID(), groupID, k.Name, k.Description, k.Measure, k.Target, k.Frequency, k.Weight, ki); err != nil {
				return err
			}
		}
	}
	return tx.Commit(ctx)
}
